### ISSUE TITLE
Form/submission authentication in local server

### ISSUE THREAD

**Issue by sjudeng** [kobotoolbox/enketo-express#487]
Hi, thank you for this open source project. I'm experimenting with a local KoBoToolbox server, using the kobo-docker project, which has made the installation process very easy (thank you again).

The only issue I'm having trouble with is the form/submission authentication. When I update user profile settings to "require authentication to see forms and submit data" and then visit my Enketo Express form, I'm prompted for kobo credentials, as expected. However, when I submit my credentials there's a page redirect and then it returns again to the credentials prompt and I'm not able to get into the data entry page. Any ideas what this might be? I'm using the latest kobo-docker, with the enketo container pointing at the latest enketo_express docker branch.

![ee-auth-prompt](https://cloud.githubusercontent.com/assets/11000076/14788080/f93a18fa-0acb-11e6-80b6-abbfe145117a.png)

**Comment by MartijnR** [comment]
Not sure what this could be, but let's leave this issue open until you find out more.

Is there an `__enketo` cookie (after entering credentials)?

Does deleting the `__enketo` cookie, then revisiting the webform, entering credentials when prompted, resolve it perhaps?

**Comment by sjudeng** [comment]
After entering credentials, there is an __enketo cookie. If I delete the cookie and then revisit the page, the cookie is regenerated. But if I purposefully enter incorrect credentials, the behavior is the same. I'm using the KOBO_SUPERUSER_USERNAME/PASSWORD.

**Comment by MartijnR** [comment]
Thanks. I am not familiar with those super user credentials. I'm guessing that should be fine, but am not sure.

To investigate further: Enketo is using the credentials to make a request to https://yourkobo.com/youraccount/formList.

You could manually try to make this request with curl to see if the page is returned, e.g. like this:

```bash
curl -u USERNAME:PASSWORD  -H 'X-OpenRosa-Version: 1.0' https://YOURKOBOSERVER.COM/USERNAME/formList
```

This will tell us whether the issue is on the KC (or credentials) or Enketo side.

**Comment by sjudeng** [comment]
Thanks, that was exactly what I needed. The issue was the server cert wasn't trusted in that container. Added an init script with the below contents and that resolved it. Thanks again.

```bash
cp /tmp/kobo_toolbox_secrets/rootCA.crt /usr/share/ca-certificates
echo "rootCA.crt" >> /etc/ca-certificates.conf
update-ca-certificates --fresh
```

**Comment by MartijnR** [comment]
Great!

**Comment by sjudeng** [comment]
Unfortunately on further testing this still won't quite go away. What's interesting is that initially it won't let me into the submission page, as described above, but then I'll execute the above curl command, which is successful, and after that I'm able to access the submission page.

**Comment by MartijnR** [comment]
ping @esmail, in case you have any suggestions for this (SSL cert?) issue between docker containers or want to move this issue elsewhere.

**Comment by sjudeng** [comment]
I found a solution by setting ```options.auth.sendImmediately=true``` in ```communicator.js```. Is this change likely to cause problems for me? Maybe curl sends credentials immediately as well, which is why that worked.

According to Node Request docs, the current default of ```sendImmediately=false``` should work as long as the server returns a 401 and includes a ```WWW-Authenticate``` header. But when auth isn't in the request, the kobocat response does appear to be correct (e.g. statusCode=401, 'www-authenticate' in the header), so I don't know what to make of that.

https://github.com/request/request#http-authentication

**Comment by esmail** [comment]
Seems like a SSL issue. @sjudeng, I'm assuming you're using a self-signed certificate and don't have a CA-signed cert. for `ee.my.domain`?

**Comment by sjudeng** [comment]
Hi @esmail, thanks for your work on these projects. Getting a local server up has been easy thanks to kobo-docker.

Regarding your question, yes, I am using a self-signed cert for *.my.domain. I have added the CA cert to the trust store on the container (enketo_express). Regarding the connection to https://kc.my.domain/user/formList, curling works from the container, and I also just found that it works in the client if I update enketo_express communicator.js to send the credentials immediately (see above). But not entirely sure of the repercussions of this change.
